apex/transformer/pipeline_parallel/schedules/fwd_bwd_pipelining_with_interleaving.py

```diff
@@ -284,6 +329,19 @@ def backward_step_helper(microbatch_id: int) -> torch.Tensor:
             grad_scaler=grad_scaler,
             deallocate_pipeline_outputs=deallocate_pipeline_outputs,
         )
+
+        # launch grad synchronization (custom grad sync)
+        # Note: To achieve maximum performance, pipeline parallelism
+        # assumes all ranks have the same compute time. However,
+        # asynchronous communication tends to slow down compute. Thus,
+        # we launch asynchronous communication at the same time across
+        # the pipeline-parallel group.
+        if custom_grad_sync_func is not None:
+            grad_sync_microbatch_id = microbatch_id - pipeline_parallel_rank
+            if grad_sync_microbatch_id >= 0 and is_last_microbatch_for_model_chunk(grad_sync_microbatch_id):
+                grad_sync_chunk_id = get_model_chunk_id(grad_sync_microbatch_id, forward=False)
+                enable_grad_sync()
+                custom_grad_sync_func(model[grad_sync_chunk_id].parameters())
```

Comment:
just to confirm: it's user's responsibility to take care of `custom_grad_sync_func` appropriately (= create a list of gradients from the list of parameters). is this correct?

Reply:
Yep, that's right. A minimal implementation would be:
```python
def grad_sync(params):
  for param in params:
    torch.distributed.all_reduce(param.grad)
```